cvmfs/cvmfs_suid_helper.cc

```diff
@@ -104,64 +105,129 @@ static void KillCvmfs(const string &fqrn) {
+  }
+
+  operator bool() const { return directory_handle_ != NULL; }
+
+  struct DirectoryEntry {
+    bool   is_directory;
+    string name;
+  };
+
+  bool NextDirectoryEntry(DirectoryEntry *entry) {
+    platform_dirent64 *dirent;
+    while ((dirent = platform_readdir(directory_handle_)) != NULL &&
+           IsMetaEntry(dirent)) {}
+    if (dirent == NULL) {
+      return false;
     }
 
     platform_stat64 info;
-    retval = platform_lstat(dirent->d_name, &info);
-    if (retval != 0) {
-      closedir(dirp);
+    if (platform_lstat(dirent->d_name, &info) != 0) {
       return false;
     }
 
-    if (S_ISDIR(info.st_mode)) {
-      // Recursion
-      retval = chdir(dirent->d_name);
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
-      retval = ClearWorkingDir();
-      if (!retval) {
-        closedir(dirp);
-        return false;
-      }
-      retval = chdir("..");
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
-      retval = rmdir(dirent->d_name);
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
+    entry->is_directory = S_ISDIR(info.st_mode);
+    entry->name         = dirent->d_name;
+    return true;
+  }
+
+ protected:
+  string GetCurrentWorkingDirectory() {
+    char path[PATH_MAX];
+    const char* cwd = getcwd(path, PATH_MAX);
+    assert(cwd == path);
+    return string(cwd);
+  }
+
+  void ChangeDirectory(const string &path) {
+    const int retval = chdir(path.c_str());
+    assert(retval == 0);
+  }
+
+  bool IsMetaEntry(const platform_dirent64 *dirent) {
+    return (strcmp(dirent->d_name, ".")  == 0) ||
+           (strcmp(dirent->d_name, "..") == 0);
+  }
+
+ private:
+  const string  previous_path_;
+        DIR    *directory_handle_;
+};
+
+static bool ClearDirectory(const string &path) {
+  ScopedWorkingDirectory swd(path);
+  if (!swd) {
+    return false;
+  }
+
+  bool success = true;
+  ScopedWorkingDirectory::DirectoryEntry dirent;
+  while (success && swd.NextDirectoryEntry(&dirent)) {
+    success = (dirent.is_directory)
+      ? ClearDirectory(dirent.name) && (rmdir(dirent.name.c_str()) == 0)
+      : (unlink(dirent.name.c_str()) == 0);
+  }
+
+  return success;
+}
+
+static int CleanupDirectory(const string &path) {
+  if (!ClearDirectory(path)) {
+    fprintf(stderr, "failed to clear %s\n", path.c_str());
+    return 1;
+  }
+  return 0;
+}
+
+static int DoSynchronousScratchCleanup(const string &fqrn) {
+  const string scratch = string(kSpoolArea) + "/" + fqrn + "/scratch/current";
+  return CleanupDirectory(scratch);
+}
+
+static int DoAsynchronousScratchCleanup(const string &fqrn) {
+  const string wastebin = string(kSpoolArea) + "/" + fqrn + "/scratch/wastebin";
+
+  // double-fork to daemonize the process
+  pid_t pid;
+  int statloc;
+  if ((pid = fork()) == 0) {
+    int retval = setsid();
+    assert(retval != -1);
+    if ((pid = fork()) == 0) {
+      close(0); close(1); close(2);
```

Comment:
[Fixed](https://github.com/cvmfs/cvmfs/pull/1541/commits/4763580be4b970d575b110a6168affc77445c1a0) by code duplication after all. I didn't factor out `Daemonize()` in `util/posix.h` because it calls to `LogCvmfs()` whose dependency I'd like to avoid.